### DISCUSSION TITLE
Is there a way to optimize Cypress launch time (terminal run for CI contexts)?

### DISCUSSION THREAD

**Discussion by sandra-ouadghiri** [cypress-io/cypress#16675] [CI setup] [answered]
Hi,

### Context
I'm in a pipeline context, doing cypress run (terminal execution).

I notice that:
- Tests take around 16s overall to execute in 6.5.0 (info available on the sum-up table last line) <- I'm happy with that
- Whole step (npx cypress run process) takes 2m53 <- I'd like to optimize that.
- Those are Azure Pipelines (yaml) ; steps does `npx cypress` run on an agent.

I'd like to optimize the CI step as much as possible, so that it takes as close to those 16s as possible.
I gave a try at 7.3 (latest) but I gained +3mn at the step level and +1m at the scenario level. So I remained willingly in v6 for performance reasons.

### Job content

```
cd tests && npx cypress run --env target=<myEnvName>
============= Starting Command Output =============
It looks like this is your first time using Cypress: 6.5.0

25l[10:45:05]  Verifying Cypress can run /home/vsts/.cache/Cypress/6.5.0/Cypress [started]
[10:45:08]  Verifying Cypress can run /home/vsts/.cache/Cypress/6.5.0/Cypress [completed]
25h
Opening Cypress...
Couldn't determine Mocha version

=============================================

  (Run Starting)

  ┌─────────────────────────┐
  │ Cypress:    6.5.0                                         │
  │ Browser:    Electron 87 (headless)             │
  │ Specs:      8 found (<list of scenarios>)    │
  └─────────────────────────┘

<Executions here>
================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
<Table>
✖  n of n failed (m%)                         00:16       <other values> 
```

My thoughts:
- I wonder what are those '25l" and "25h" at the job beginning. I saw that on other people terminal executions too (ex on stackoverflow when they share). Not sure if this means anything in regards to my concern.
- "It looks like this is your first time using Cypress: 6.5.0" > Not sure if this makes me loose time (does it add /example folder for ex?) nor if this could be improved/prevented. I'm on an Ubuntu azure agent. As far as possible, we'd like to stay on a built-in agent to prevent sustenance. But this is open for discussion obviously.
- "Couldn't determine Mocha version" is [issue 9439](https://github.com/cypress-io/cypress/issues/9439)
- I have modifications on plugins/index.js ; mainly linked to us using Cucumber.

I'd like a way to identify what takes time (where 90% of the step time goes) and/or any advise on how I can reduce that fat around tests execution time.

I need to divide my tests into 2 CI steps, and I worry this will add extra load. Cypress seems slow to start and this gives a bad experience to our developpers in our CI context. Any feedback on how to mitigate or optimize that is very welcome.

**Comment by jennifer-shehane** [comment, marked as answer]
You can see more timings with [DEBUG logs](https://on.cypress.io/troubleshooting#Print-DEBUG-logs) on.

That message about verifying Cypress essentially means it will run `cypress verify` command to ensure everything is ok, dependency-wise, to run. 

You're likely seeing an increase in timing in 7.0 due to the screen size increase (this is what I've identified). You could try reducing the screen size to the default size prior to 7.0 (1280x720) to see if the increased time is related to the increase in screen size. [Instructions here.](https://on.cypress.io/browser-launch-api#Set-screen-size-when-running-headless)

I think the `25l` and `25h` are part of some malformed ANSI characters. It can be ignored. 

Generally I've seen people's runtimes increase with whatever file preprocessing they're doing which would run before cypress opens. Not sure if you have any of that setup in your plugins file.

**Reply by san-slysz** [reply]
Thanks @jennifer-shehane 
I actually added some code here, so your feedback makes total sense.
I will dig using the debug logs to learn time is lost just to see if everything is as optimised as possible.
